demo/site/src/middleware/predefinedPages.ts

```diff
@@ -0,0 +1,77 @@
+import { gql } from "@comet/cms-site";
+import { languages } from "@src/config";
+import { predefinedPagePaths } from "@src/predefinedPages/predefinedPagePaths";
+import { createGraphQLFetch } from "@src/util/graphQLClient";
+
+import { memoryCache } from "./cache";
+import { GQLPredefinedPagesQuery, GQLPredefinedPagesQueryVariables } from "./predefinedPages.generated";
+
+async function getPredefinedPageRedirect(scope: { domain: string }, pathname: string): Promise<string | undefined> {
+    const pages = await fetchPredefinedPages(scope);
+
+    const matchingPredefinedPage = pages.find((page) => pathname.startsWith(page.defaultPath));
+
+    if (matchingPredefinedPage) {
+        return pathname.replace(matchingPredefinedPage.defaultPath, matchingPredefinedPage.customPath);
+    }
+
+    return undefined;
+}
+
+async function getPredefinedPageRewrite(scope: { domain: string }, pathname: string): Promise<string | undefined> {
+    const pages = await fetchPredefinedPages(scope);
+
+    const matchingPredefinedPage = pages.find((page) => pathname.startsWith(page.customPath));
+
+    if (matchingPredefinedPage) {
+        return pathname.replace(matchingPredefinedPage.customPath, matchingPredefinedPage.defaultPath);
+    }
+
+    return undefined;
+}
+
+const predefinedPagesQuery = gql`
+    query PredefinedPages($scope: PageTreeNodeScopeInput!) {
+        paginatedPageTreeNodes(scope: $scope, documentType: "PredefinedPage") {
+            nodes {
+                id
+                path
+                document {
+                    __typename
+                    ... on PredefinedPage {
+                        type
+                    }
+                }
+            }
+        }
+    }
+`;
+
+const graphQLFetch = createGraphQLFetch();
+
+async function fetchPredefinedPages(scope: { domain: string }) {
+    const key = `predefinedPages-${JSON.stringify(scope)}`;
+
+    const pages: Array<{ defaultPath: string; customPath: string }> = [];
```

Comment:
naming is not easy to understand imho, maybe:

- defaultPath -> codePath
- customPath -> configuredPath

(or you have a better idea)

Reply:
Changed to codePath and pageTreeNodePath: 2011f235db8dcf397b0f4aa9a581886f636535a6